### PR TITLE
Add support for expressions within template literals

### LINKED ISSUE #25: Expressions within template literal don't work
I would understand if dynamic properties didn't work, but dynamic values don't seem to be working either. Compiling succeeds, but the pieces with the expression seem to be omitted from the inserted script tag.

```
<div>
	<div id={this.state.calendarId}></div>
	<style jsx>{`
			#${this.state.calendarId} .fc-button.fc-state-default {
				border: 2px solid ${buttonColor};
			}
	`}</style>
</div>
```
image

---

```
const buttonColor = 'red';

return (
	<div>
		<div id="foo"></div>
		<style jsx>{`
			#foo .fc-button.fc-state-default {
				border: 2px solid ${buttonColor};
			}
		`}</style>
	</div>
);
```
image

---

```
<div>
	<div id="foo"></div>
	<style jsx>{`
		#foo .fc-button.fc-state-default {
			border: 2px solid red;
		}
	`}</style>
</div>
```
image

Thank you!

### PR DESCRIPTION
**Pull request by giuseppeg**
Fixes #25

### CODE REVIEW

**Comment by giuseppeg** [comment]
fyi this patch doesn't allow expressions in the selector name like `.col-${xsmall}` not sure if that's necessary or that useful.

**Comment by sarukuku** [comment]
Waiting for this to land. 🚀

**Comment by giuseppeg** [comment]
@rauchg, @nkzawa I think that this branch is ready for review.
Let me know if you need context but hopefully this solution makes sense and it is not too dumb :)

File: src/babel.js [hunk 1]
```diff
@@ -17,23 +18,98 @@ const STYLE_COMPONENT_CSS = 'css'
 export default function ({types: t}) {
   const findStyles = children => (
     children.filter(el => (
```
Comment (timneutkens) [inline src/babel.js]: Just a small thing, think you can destructure this param `({node}) => (` which allows you to remove `el` in front of `node`

Comment (giuseppeg) [inline src/babel.js]: oh right good catch!

Comment (timneutkens) [inline src/babel.js]: :fire:

**Comment by rauchg** [comment]
Are we erroring upon usage of vars from the closure?

**Comment by giuseppeg** [comment]
No but we could although it might add some overhead or be not that straightforward if we keep allowing things like call expressions

**Comment by giuseppeg** [comment]
basically you want to make sure that people don't use `props` right?

**Comment by rauchg** [comment]
Correct. If this doesn't fail it will create extremely difficult-to-understand bugs to the uninitiated:

```jsx
export default ({ some, props}) => {
  <div>
    <style jsx>{`
      div { color: ${some} }
    `}</style>
  </div>
}
```

**Comment by giuseppeg** [comment]
@rauchg any identifier in the scope or just props?

```jsx
export default () => {
  const some = 'red' // allowed or not?

  return (<div>
    <style jsx>{`
      div { color: ${some} }
    `}</style>
  </div>)
}
```

**Comment by rauchg** [comment]
I'd block anything from the scope. After all, if it's a constant, you should probably syntactically move it outside the scope…

File: src/babel.js [hunk 2]
```diff
@@ -16,24 +17,120 @@ const STYLE_COMPONENT_CSS = 'css'
...
+  }
 
-  const makeStyledJsxTag = (id, transformedCss) => (
-    t.JSXElement(
+  const makeStyledJsxTag = (id, transformedCss, isTemplateLiteral) => {
+    let css
+    if (isTemplateLiteral) {
+      // build the expression from transformedCss
+      traverse(
+        parse(`\`${transformedCss}\``).ast,
```
Comment (nkzawa) [inline src/babel.js]: I think we should use https://github.com/babel/babylon

Comment (giuseppeg) [inline src/babel.js]: babel-core's transform uses babylon behind the wood

Comment (nkzawa) [inline src/babel.js]: yep, but I meant it would be better to use the parser anyway.

Comment (sedubois) [inline src/babel.js]: Behind the wood 😄👍

Comment (giuseppeg) [inline src/babel.js]: rofl typo. @nkzawa yea sorry I was rushing. In fact I thought of doing that initially but then I didn't feel like adding the extra dependency but if you think that it is preferable I have no problem with that.

**Comment by rauchg** [comment]
Needs rebase

**Comment by giuseppeg** [comment]
@rauchg you got it

**Comment by sarukuku** [comment]
Any updates on this? 🚀

**Comment by giuseppeg** [comment]
@sarukuku I haven't had a lot of free time recently. Anyway we want to benchmark this before merging – I will try to push a minimal benchmark today after work.

**Comment by giuseppeg** [comment]
301 to #80 :)

**Pull request closed without merging**